hanish/zipcode.py

```diff
@@ -0,0 +1,168 @@
+
+
+SCHEMA = (
+    "CREATE TABLE IF NOT EXISTS zipcodes ("
+        "zipcode TEXT, "
+        "latitude REAL, "
+        "longitude REAL"
+    ")"
+)
+
+
+##########################################################################
+## ZipCodeDB
+##########################################################################
+
+class ZipCodeDB(object):
+    """
+    A ZipCodeDB is loaded from a CSV file that maps US Zip Codes to latitutde,
+    longitude coordinates. The data is loaded into an in-memory Sqlite3
+    database for quick querying. The data file can be downloaded from:
+
+        https://gist.github.com/erichurst/7882666
+
+    It is also included in the fixtures directory of the repository.
+
+    Usage:
+
+        # Load the db and lookup coordinates
+        zipdb = ZipCodeDB.load('fixtures/ziplatlon.csv')
+        lat, lon = zipdb.lookup('20001')
+
+        # When done, close the db
+        zipdb.close()
+
+    Note: Zip Codes are strings, not numbers (becasuse of leading zeros).
+    """
+
+    @classmethod
+    def load(klass, path):
+        """
+        Load the zip code database from a CSV file that has three columns:
+
+            ZIP,LAT,LNG
+
+        If the database is not initialized with load, then it will be empty.
+        """
+
+        # Initialize the database and get a cursor
+        db = klass()
+        cursor = db.conn.cursor()
+
+        # Read the data from the CSV file and insert into db
+        with open(path, 'r') as f:
+            reader = csv.DictReader(f)
+            for row in reader:
+                args = (row['ZIP'], float(row['LAT']), float(row['LNG']))
+                cursor.execute(
+                    "INSERT INTO zipcodes VALUES (?,?,?)", args
+                )
+
+        # Commit the database operations
+        db.conn.commit()
+
+        # Return the loaded database
+        return db
+
+    def __init__(self):
+        # Connect to the database
+        self.conn = sqlite3.connect(":memory:")
+
+        # Create the table(s) from the schema
+        cursor = self.conn.cursor()
+        cursor.execute(SCHEMA)
+        self.conn.commit()
+
+    def close(self):
+        """
+        Close the connection to the database, removing data from memory and
+        allowing no further accesses (otherwise an exception will be raised).
+        """
+        self.conn.close()
+        self.conn = None
+
+    def execute(self, sql, *args, **kwargs):
+        """
+        Helper function that creates a cursor and executes a single SQL
+        statement, returning the cursor to read the results of the query.
+        """
+        # Check the connection
+        if self.conn is None:
+            raise DatabaseError(
+                "Zip Code database has been closed and removed from memory"
+            )
+
+        # Execute the query and return
+        cursor = self.conn.cursor()
+        cursor.execute(sql, *args, **kwargs)
+        return cursor
+
+
+    def lookup(self, zipcode):
```

Comment:
your docs state that zip codes need to be limited to 5-digit strings and you gracefully handle not finding a match, but this does highlight the interesting case of "full length" US postal codes "94110-1234"

can you opine briefly on where we should draw the line for this in a dynamic language? (should our method try and be clever and truncate the input if its too long, should we use the same error for a missing value in the database as for a malformed argument, etc)